File: Matrix_Multiplication_Python/MatrixMultiplicationTest.py
import time
import gc
from memory_profiler import memory_usage
from MatrixMultiplication import MatrixMultiplication

class MatrixMultiplicationTest:
    def __init__(self, matrix_sizes, runs_per_size=5):
        self.matrix_sizes = matrix_sizes
        self.runs_per_size = runs_per_size

    def get_used_memory(self):
        mem = memory_usage(-1, interval=0.1, timeout=1)
        return mem[0]

    def run_test(self):
        for size in self.matrix_sizes:
            print(f"=== Tamaño de Matriz: {size}x{size} ===")
            total_time = 0.0
            total_memory = 0.0

            for run in range(1, self.runs_per_size + 1):
                mm = MatrixMultiplication(size)
                mm.initialize_matrices()
                memory_before = self.get_used_memory()
                start_time = time.time()
                mm.multiply()
                end_time = time.time()
                memory_after = self.get_used_memory()
                elapsed_time = end_time - start_time
                total_time += elapsed_time
                memory_used = memory_after - memory_before
                total_memory += memory_used
                print(f"Ejecución {run}: Tiempo = {elapsed_time:.3f} segundos, Memoria Usada = {memory_used:.2f} MB")

            average_time = total_time / self.runs_per_size
            average_memory = total_memory / self.runs_per_size
            print(f"Promedio: Tiempo = {average_time:.3f} segundos, Memoria Promedio Usada = {average_memory:.2f} MB\n")

if __name__ == "__main__":
    matrix_sizes = [256, 512, 1024, 2048]
    runs_per_size = 5

    tester = MatrixMultiplicationTest(matrix_sizes, runs_per_size)
    tester.run_test()

File: MatrixMultiplicationJava/src/test/java/MatrixMultiplicationTest.java
public class MatrixMultiplicationTest {

    public static void main(String[] args) {
        int[] matrixSizes = {256, 512, 1024, 2048};
        int runsPerSize = 5;

        for (int size : matrixSizes) {
            System.out.println("=== Tamaño de Matriz: " + size + "x" + size + " ===");
            double totalTime = 0.0;
            long totalMemory = 0;

            for (int run = 1; run <= runsPerSize; run++) {
                MatrixMultiplication mm = new MatrixMultiplication(size);
                mm.initializeMatrices();
                long memoryBefore = getUsedMemory();
                long startTime = System.nanoTime();
                mm.multiply();
                long endTime = System.nanoTime();
                long memoryAfter = getUsedMemory();
                double elapsedTime = (endTime - startTime) / 1e9;
                totalTime += elapsedTime;
                long memoryUsed = (memoryAfter - memoryBefore) / (1024 * 1024);
                totalMemory += memoryUsed;
                System.out.printf("Ejecución %d: Tiempo = %.3f segundos, Memoria Usada = %d MB%n",
                        run, elapsedTime, memoryUsed);
            }
            double averageTime = totalTime / runsPerSize;
            double averageMemory = (double) totalMemory / runsPerSize;
            System.out.printf("Promedio: Tiempo = %.3f segundos, Memoria Promedio Usada = %.2f MB%n%n",
                    averageTime, averageMemory);
        }
    }

    private static long getUsedMemory() {
        Runtime runtime = Runtime.getRuntime();
        return runtime.totalMemory() - runtime.freeMemory();
    }
}
